modules/gapi/src/streaming/gstreamer/gstreamer_pipeline_facade.cpp

```diff
@@ -0,0 +1,329 @@
+
+    if (outElements.empty())
+    {
+        return nullptr;
+    }
+    else
+    {
+        GAPI_Assert(1ul == outElements.size());
+        return outElements[0];
+    }
+}
+
+void GStreamerPipelineFacade::completePreroll() {
+    // FIXME: If there are multiple sources in pipeline and one of them is live, then pipeline
+    //        will return GST_STATE_CHANGE_NO_PREROLL while pipeline pausing.
+    //        But appsink may not be connected to this live source and only to anothers,
+    //        not-live ones. So, it is not required to start the playback for appsink to complete
+    //        the preroll.
+    //        Starting of playback for the not-live sources before the first frame pull will lead
+    //        to loosing of some amount of frames and pulling of the first frame can return frame
+    //        which is far from the first.
+    //
+    //        Need to handle this case or forbid to mix multiples sources of different
+    //        categories(live, not-live) in the pipeline explicitly(assert).
+
+    if (!m_isPrerolled.load(std::memory_order_acquire))
+    {
+        std::lock_guard<std::mutex> lock(m_mutex);
+
+        if(!m_isPrerolled.load(std::memory_order_relaxed))
+        {
+            PipelineState state = queryState();
+
+            // Only move forward in the pipeline's state machine
+            GAPI_Assert(state.current != GST_STATE_PLAYING);
+
+            GAPI_Assert(state.pending == GST_STATE_VOID_PENDING);
+            GstStateChangeReturn status = gst_element_set_state(m_pipeline, GST_STATE_PAUSED);
+            checkBusMessages();
+            if (status == GST_STATE_CHANGE_NO_PREROLL)
+            {
+                status = gst_element_set_state(m_pipeline, GST_STATE_PLAYING);
+                m_isPlaying.store(true);
+            }
+            verifyStateChange(status);
+
+            m_isPrerolled.store(true, std::memory_order_release);
+        }
+    }
+}
+
+void GStreamerPipelineFacade::play()
+{
+    if (!m_isPlaying.load(std::memory_order_acquire))
+    {
+        std::lock_guard<std::mutex> lock(m_mutex);
+
+        if (!m_isPlaying.load(std::memory_order_relaxed))
+        {
+            setState(GST_STATE_PLAYING);
+            m_isPlaying.store(true, std::memory_order_release);
+            m_isPrerolled.store(true);
+        }
+    }
+}
+
+bool GStreamerPipelineFacade::isPlaying() {
+    return m_isPlaying.load();
+}
+
+std::vector<GstElement*> GStreamerPipelineFacade::getElements(
+    std::function<bool(GstElement*)> comparator)
+{
+    std::vector<GstElement*> outElements;
+    GStreamerPtr<GstIterator> it(gst_bin_iterate_elements(GST_BIN(m_pipeline.get())) /* transfer full ???*/);
+
+    gboolean done = false;
+    GValue value = G_VALUE_INIT;
+    while (!done)
+    {
+        switch (gst_iterator_next(it, &value))
+        {
+            case GST_ITERATOR_OK:
+            {
+                GstElement* element = GST_ELEMENT(g_value_get_object(&value)); /* transfer none */
```

Comment:
is there any guarantee from gstreamer  API about `element` is no dangling pointer after `value` has gone?
I means end up being GstElement* valid after function exit? or need make a copy?